### ISSUE TITLE
Librispeech Question

### ISSUE THREAD

**Issue by dsa934** [espnet/espnet#1175]
hello, I am studying librispeech recipe and have 2 questions.

Q1. 
Why use the Average ASR / LM model in the decoding stage ?
Does it mean that we store a snapshot of the appropriate weights found during learning in stage 4 and use the best of them in the decoding stage?

Q2.
How can I use other NN models?
In stage4, it seems that the encoder-decoder model is composed of BLSTM.

If i want to experiment with models such as li-gru .. then
Do I need to link after generating a file like mymodel.py? If so, how do i connect?
Or is there another way?

**Comment by b-flo** [comment]
Hi,

Q1: During training, a snapshot is created at each validation step. The best model is selected according to a criterion (loss, accuracy), for each validation step if the value is lower (or higher) compared to the previous step the best model is updated/replaced.

Q2: The architecture of your model is defined by the config file you use. Take a look at the available configs file for different recipes (`egs/[recipe]/conf/`) to have an idea.
Also I don't know if there is an easier method, but you could also take a look at `https://github.com/espnet/espnet/tree/master/espnet/nets/pytorch_backend/e2e_[type_of_model].py`, it contains the available choices for each part of the architecture.

**Comment by sw005320** [comment]
Additional comments to @b-flo's answers

Q1: averaging is quite powerful for Transformer (or I could say the warm-up optimizer). On the other hand, RNN/LSTM with adadelta does not have so large gains. Anyway, if you use Transformer, you should use the averaged model.

**Comment by dsa934** [comment]
https://github.com/espnet/espnet/blob/2c961b3cd1843b332f5d5f960083abe3235ce0e5/espnet/bin/asr_train.py#L188

Can you explain about WPE?
I tried changing the configuration of the encoder decoder model from blstm to ligru but it didn't work.

If I run the code after modifying it like this:
https://github.com/espnet/espnet/blob/2c961b3cd1843b332f5d5f960083abe3235ce0e5/espnet/bin/asr_train.py#L189
default = 'gru' then 
Is it correct that the encoder-decoder network is composed of gru?

**Comment by MarkWuNLP** [comment]
> Additional comments to @b-flo's answers
> 
> Q1: averaging is quite powerful for Transformer (or I could say the warm-up optimizer). On the other hand, RNN/LSTM with adadelta does not have so large gains. Anyway, if you use Transformer, you should use the averaged model.

It sounds very interesting. @sw005320 Do you remember the WER improvement on Librispeech dataset with the averaging checkpoints?

Thank you!
Yu

**Comment by sw005320** [comment]
> It sounds very interesting. @sw005320 Do you remember the WER improvement on Librispeech dataset with the averaging checkpoints?

I don't fully record it, but usually, we'll get 10% relative improvement.
